autotest/alg/contour.py

```diff
@@ -169,40 +162,31 @@ def contour_2():
 
     lyr = ogr_ds.ExecuteSQL("select * from contour order by elev asc")
 
-    if lyr.GetFeatureCount() != len(expected_envelopes):
-        print('Got %d features. Expected %d' % (lyr.GetFeatureCount(), len(expected_envelopes)))
-        return 'fail'
+    assert lyr.GetFeatureCount() == len(expected_envelopes), ('Got %d features. Expected %d' % (lyr.GetFeatureCount(), len(expected_envelopes)))
 
     i = 0
     feat = lyr.GetNextFeature()
     while feat is not None:
-        if feat.GetGeometryRef().GetZ(0) != expected_height[i]:
-            print('Got %f as z. Expected %f' % (feat.GetGeometryRef().GetZ(0), expected_height[i]))
-            return 'fail'
+        assert feat.GetGeometryRef().GetZ(0) == expected_height[i], ('Got %f as z. Expected %f' % (feat.GetGeometryRef().GetZ(0), expected_height[i]))
         envelope = feat.GetGeometryRef().GetEnvelope()
-        if feat.GetField('elev') != expected_height[i]:
-            print('Got %f. Expected %f' % (feat.GetField('elev'), expected_height[i]))
-            return 'fail'
+        assert feat.GetField('elev') == expected_height[i], ('Got %f. Expected %f' % (feat.GetField('elev'), expected_height[i]))
         for j in range(4):
             if abs(expected_envelopes[i][j] - envelope[j]) > precision / 2 * 1.001:
-                print('i=%d, wkt=%s' % (i, feat.GetGeometryRef().ExportToWkt()))
                 print(feat.GetGeometryRef().GetEnvelope())
                 print(expected_envelopes[i])
                 print('%f, %f' % (expected_envelopes[i][j] - envelope[j], precision / 2))
-                return 'fail'
+                pytest.fail('i=%d, wkt=%s' % (i, feat.GetGeometryRef().ExportToWkt()))
```

Comment:
Can this type of case be turned into an assert with a message rather than an if and a bunch of prints with a `pytest.fail` at the end?  This is where it shows that I spend 99% of my python testing time working with `unittest` and know almost no pytest.

This is a case where I would write this with unittest: ( and I worry about that integer 2 literal... sigh)

    self.assertAlmostEqual(expected_envelopes[i][j], envelope[j], delta=precision / 2.0 * 1.001, msg = ...)

Reply:
Agreed. Leave the output up to pytest and remove the print statements.

Reply:
Is that generally applicable though? I feel the print calls can sometimes be useful, since pytest only dumps expressions in the condition itself, and sometimes more information could be handy for debugging

Reply:
```python
self.assertAlmostEqual(expected_envelopes[i][j], envelope[j], delta=precision / 2.0 * 1.001, msg = ...)
```

pytest has an equivalent:

```python
assert expected_envelopes[i][j] == pytest.approx(envelope[j], abs=precision / 2 * 1.001)
```

Reply:
If the test passes, you don't need more information, right? If you do, then the test needs to be broken up.

Pytest is smart about capturing log output from failed tests, so the thing to do is use the logging module, not `print()`.

Personally, I think that use of logging (or print) for output from a test is a sign that the test needs to be rewritten to make more asserts or be split into new tests that assert different things.

Reply:
I agree with you. However, the only way that I can apply that (in an automated way, to the entire gdal test suite) is to drop _all_ print statements in these `if ... pytest.fail()` blocks.

Doing that would actually make things simpler, but I anticipated pushback on it. I'm happy to strip them out 😄

Reply:
Okay, I'm going to do this


```diff
-    if cs != 0:
-        print(cs)
-        return 'fail'
+    assert cs == 0
```

If the only names in the print/post_reason call are also part of the expression being tested, I'll throw them away (pytest will output them anyway when the assertion fails)

However:

```diff
-        if out_ds is not None:
-            print(i)
-            return 'fail'
+        assert out_ds is None, i
```

If there's something being print/post_reason'd which _isn't_ in the expression being tested, I'll keep it for the reason part of the assertion.

I'll strip all print/post_reason calls regardless.

Reply:
For clarification, Pytest captures `print` output the same way it captures log messages, so it’s only visible in the console for failing tests.

Reply:
> And it’ll never get finished if we need to refactor fully grimacing

I agree that improvements on error messages can be a later effort. The priority is to make sure that the refactoring has the same success / error status than the current test suite, and make the CI happy (as CI doesn't have all the drivers built in, in particular for proprietary ones, I'd be happy to check again on a local build)

Reply:
Okay. I've got it doing a sensible halfway solution.

It now finds a reason from `post_reason` and `print` statements. It discards all the ones it detects as low value, as well as the one it selects as the reason. If there are any print statements left over, it keeps them and produces a `pytest.fail` statement. Otherwise it produces an `assert`. 

I've also added a regex for 'got X, expected Y' strings to mostly-detect pointless prints.

The result is, almost all the time it produces an `assert`, but occasionally it still produces a `pytest.fail`.


Some examples:

```diff
-    if lyr.GetFeatureCount() != len(expected_envelopes):
-        print('Got %d features. Expected %d' % (lyr.GetFeatureCount(), len(expected_envelopes)))
-        return 'fail'
+    assert lyr.GetFeatureCount() == len(expected_envelopes)
```

```diff
             if abs(expected_envelopes[i][j] - envelope[j]) > precision / 2 * 1.001:
                 print('i=%d, wkt=%s' % (i, feat.GetGeometryRef().ExportToWkt()))
                 print(feat.GetGeometryRef().GetEnvelope())
-                print(expected_envelopes[i])
-                print('%f, %f' % (expected_envelopes[i][j] - envelope[j], precision / 2))
-                return 'fail'
+                pytest.fail('%f, %f' % (expected_envelopes[i][j] - envelope[j], precision / 2))
```